capsules/src/sht3x.rs

```diff
@@ -0,0 +1,308 @@
+//!
+//!
+
+use core::cell::Cell;
+use enum_primitive::cast::FromPrimitive;
+use enum_primitive::enum_from_primitive;
+use kernel::common::cells::{OptionalCell, TakeCell};
+use kernel::hil::i2c;
+use kernel::hil::time::{self, Alarm};
+use kernel::ReturnCode;
+
+pub static BASE_ADDR: u8 = 0x44;
+
+#[repr(u16)]
+enum_from_primitive! {
+    enum Registers {
+        /// Measurement High Repeatability with Clock Stretch Enabled
+        MEASHIGHREPSTRETCH = 0x2C06,
+        /// Measurement Medium Repeatability with Clock Stretch Enabled
+        MEASMEDREPSTRETCH = 0x2C0D,
+        /// Measurement Low Repeatability with Clock Stretch Enabled
+        MEASLOWREPSTRETCH = 0x2C10,
+        /// Measurement High Repeatability with Clock Stretch Disabled
+        MEASHIGHREP = 0x2400,
+        /// Measurement Medium Repeatability with Clock Stretch Disabled
+        MEASMEDREP = 0x240B,
+        /// Measurement Low Repeatability with Clock Stretch Disabled
+        MEASLOWREP = 0x2416,
+        /// Read Out of Status Register
+        READSTATUS = 0xF32D,
+        /// Clear Status
+        CLEARSTATUS = 0x3041,
+        /// Soft Reset
+        SOFTRESET = 0x30A2,
+        /// Heater Enable
+        HEATEREN = 0x306D,
+        /// Heater Disable
+        HEATERDIS = 0x3066,
+        /// Status Register Heater Bit
+        REGHEATERBIT = 0x0d,
+    }
+}
+
+#[derive(Clone, Copy, PartialEq)]
+enum State {
+    Idle,
+    ReadStatus,
+    Read,
+    ReadData,
+    Reset,
+}
+
+fn crc8(data: &[u8]) -> u8 {
+    let polynomial = 0x31;
+    let mut crc = 0xff;
+
+    for x in 0..data.len() {
+        crc ^= data[x as usize] as u8;
+        for _i in 0..8 {
+            if (crc & 0x80) != 0 {
+                crc = crc << 1 ^ polynomial;
+            } else {
+                crc = crc << 1;
+            }
+        }
+    }
+    crc
+}
+
+pub struct SHT3x<'a, A: Alarm<'a>> {
+    i2c: &'a dyn i2c::I2CDevice,
+    humidity_client: OptionalCell<&'a dyn kernel::hil::sensors::HumidityClient>,
+    temperature_client: OptionalCell<&'a dyn kernel::hil::sensors::TemperatureClient>,
+    state: Cell<State>,
+    buffer: TakeCell<'static, [u8]>,
+    read_temp: Cell<bool>,
+    read_hum: Cell<bool>,
+    alarm: &'a A,
+}
+
+impl<'a, A: Alarm<'a>> SHT3x<'a, A> {
+    pub fn new(
+        i2c: &'a dyn i2c::I2CDevice,
+        buffer: &'static mut [u8],
+        alarm: &'a A,
+    ) -> SHT3x<'a, A> {
+        SHT3x {
+            i2c: i2c,
+            humidity_client: OptionalCell::empty(),
+            temperature_client: OptionalCell::empty(),
+            state: Cell::new(State::Idle),
+            buffer: TakeCell::new(buffer),
+            read_temp: Cell::new(false),
+            read_hum: Cell::new(false),
+            alarm: alarm,
+        }
+    }
+
+    pub fn clear_status(&self) {
+        self.buffer.take().map(|buffer| {
+            self.i2c.enable();
+
+            buffer[0] = ((Registers::CLEARSTATUS as u16) >> 8) as u8;
```

Comment:
This could probably made into a

```rust
let [high, low] = u16::to_be_bytes(Registers::CLEARSTATUS);
```

which I find more expressive, but either way is fine. :+1: